ndt7/server/upload/upload.go

```diff
@@ -0,0 +1,221 @@
+	defaultTimeout = 7 * time.Second
+
+	// defaultDuration is the default duration of a subtest in nanoseconds.
+	defaultDuration = 10 * time.Second
+)
+
+// Handler handles an upload test on the server side.
+type Handler struct {
+	Upgrader websocket.Upgrader
+	DataDir  string
+}
+
+// warnAndClose emits a warning |message| and then closes the HTTP connection
+// using the |writer| http.ResponseWriter.
+func warnAndClose(writer http.ResponseWriter, message string) {
+	logging.Logger.Warn(message)
+	writer.Header().Set("Connection", "Close")
+	writer.WriteHeader(http.StatusBadRequest)
+}
+
+// getConnFileAndPossiblyEnableBBR returns the connection to be used to
+// gather low level stats and possibly enables BBR. It returns a file to
+// use to gather BBR/TCP_INFO stats on success, an error on failure.
+func getConnFileAndPossiblyEnableBBR(conn *websocket.Conn) (*os.File, error) {
+	fp := fdcache.GetAndForgetFile(conn.UnderlyingConn())
+	// Implementation note: in theory fp SHOULD always be non-nil because
+	// now we always register the fp bound to a net.TCPConn. However, in
+	// some weird cases it MAY happen that the cache pruning mechanism will
+	// remove the fp BEFORE we can steal it. In case we cannot get a file
+	// we just abort the test, as this should not happen (TM).
+	if fp == nil {
+		err := errors.New("cannot get file bound to websocket conn")
+		logging.Logger.WithError(err).Warn("Cannot enable BBR")
+		return nil, err
+	}
+	err := bbr.Enable(fp)
+	if err != nil {
+		logging.Logger.WithError(err).Warn("Cannot enable BBR")
+		// FALLTHROUGH
+	}
+	return fp, nil
+}
+
+// gatherAndSaveTCPInfoAndBBRInfo gathers TCP info and BBR measurements from
+// |sockfp| and stores them into the |measurement| object as well as into the
+// |resultfp| file. Returns an error on failure and nil in case of success.
+func gatherAndSaveTCPInfoAndBBRInfo(measurement *model.Measurement, sockfp *os.File, resultfp *results.File) error {
+	bbrinfo, err := bbr.GetMaxBandwidthAndMinRTT(sockfp)
+	if err == nil {
+		measurement.BBRInfo = &bbrinfo
+	}
+
+	metrics, err := tcpinfox.GetTCPInfo(sockfp)
+	if err == nil {
+		measurement.TCPInfo = &metrics
+	}
+
+	if err := resultfp.WriteMeasurement(*measurement, "server"); err != nil {
+		logging.Logger.WithError(err).Warn("Cannot save measurement on disk")
+		return err
+	}
+	return nil
+}
+
+// measuringLoop is the loop that runs the measurements in a goroutine.
+// This function exits when
+//     (1) a fatal error occurs or
+//     (2) the maximum elapsed time for the upload test expires.
+// Since this function has access to BBR stats (if available), it's the
+// right place to stop the test early. The rest of the upload code is supposed
+// to stop receiving when this function signals that we're done by closing
+// the channel.
+func measuringLoop(ctx context.Context, request *http.Request,
+	conn *websocket.Conn, dataDir string, dst chan model.Measurement) {
```

Comment:
My eyes bleed because of the default `go fmt` indentation when splitting lines. (Just a random comment; when we'll refactor we will probably want to have less params here, that would be the proper fix I guess.)

Reply:
That would definitely be the proper fix.